Firestore/core/src/remote/remote_event.cc

```diff
@@ -235,15 +237,68 @@ void WatchChangeAggregator::HandleExistenceFilter(
     } else {
       int current_size = GetCurrentDocumentCountForTarget(target_id);
       if (current_size != expected_count) {
-        // Existence filter mismatch: We reset the mapping and raise a new
-        // snapshot with `isFromCache:true`.
-        ResetTarget(target_id);
-        pending_target_resets_.insert(target_id);
+        // Apply bloom filter to identify and mark removed documents.
+        bool bloom_filter_applied =
+            ApplyBloomFilter(existence_filter, current_size);
+        std::cout << "bloom_filter_applied" << bloom_filter_applied
+                  << std::endl;
+        if (!bloom_filter_applied) {
+          // If bloom filter application fails, we reset the mapping and
+          // trigger re-run of the query.
+          ResetTarget(target_id);
+          pending_target_resets_.insert(target_id);
+        }
       }
     }
   }
 }
 
+/** Returns whether a bloom filter removed the deleted documents successfully.
+ */
+bool WatchChangeAggregator::ApplyBloomFilter(
+    ExistenceFilterWatchChange existence_filter, int current_count) {
+  int expected_count = existence_filter.filter().count();
+  absl::optional<BloomFilter> bloom_filter =
```

Comment:
should I move this? This should be the only place that actually uses bloom_filter in existence_filter, so stealing it should not create a problem, unless, in the future, we have decided to re-run bloom filter application or use it somewhere else.